src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +280,144 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
+     * @param successfulSearchShardIndices the set of indices that were successfully searched
+     * @param showFieldName            whether to include field names
+     * @param showFieldType            whether to include field types
+     * @return Field data string
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    String buildFieldDataString(
+        NamedWriteable builder,
+        Set<Index> successfulSearchShardIndices,
+        Boolean showFieldName,
+        Boolean showFieldType
+    ) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            if (showFieldName) {
+                fieldDataList.add(fieldName);
+            }
+            if (showFieldType) {
+                String fieldType = getFieldType(fieldName, successfulSearchShardIndices);
+                if (fieldType != null && !fieldType.isEmpty()) {
+                    fieldDataList.add(fieldType);
+                }
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Set<Index> successfulSearchShardIndices) {
+        if (successfulSearchShardIndices == null) {
+            return null;
+        }
+        String fieldType = getFieldTypeFromCache(fieldName, successfulSearchShardIndices);
+        if (fieldType != null) {
+            if (fieldType.equals(NO_FIELD_TYPE_VALUE)) {
+                return null;
+            } else {
+                return fieldType;
+            }
+        }
+
+        Map<String, MappingMetadata> allIndicesMap = getStringMappingMetadataMap(successfulSearchShardIndices);
+        if (allIndicesMap == null) return null;
+
+        return findFieldTypeInMappings(fieldName, successfulSearchShardIndices, allIndicesMap);
+    }
+
+    private String findFieldTypeInMappings(
+        String fieldName,
+        Set<Index> successfulSearchShardIndices,
+        Map<String, MappingMetadata> allIndicesMap
+    ) {
+        for (Index index : successfulSearchShardIndices) {
+            String fieldType = getFieldTypeFromMapping(index, fieldName, allIndicesMap.get(index.getName()));
+            if (fieldType != null) {
+                cacheFieldType(index, fieldName, fieldType);
+                return fieldType;
+            }
+        }
+        return null;
+    }
+
+    String getFieldTypeFromMapping(Index index, String fieldName, MappingMetadata mappingMetadata) {
+        // Get properties from cache or mapping metadata
+        Map<String, ?> propertiesAsMap = propertiesCache.computeIfAbsent(index, k -> {
+            if (mappingMetadata != null) {
+                return (Map<String, ?>) mappingMetadata.getSourceAsMap().get("properties");
```

Comment:
I believe this should be single recursive method combined with `findFieldTypeInProperties`. It can have signature very similar to `String visitMapping(Map<String, ?> propertiesAsMap, String[] fieldName, int idx)`.

Reply:
Also, we are not handling the multi fields case.  Can we account for that as well?

Reply:
Handled multi-fields in the latest